### PR TITLE
[MGR] Solving #72 ignore mode feature

### PR DESCRIPTION
**Pull request by cohenjer**
This is a remake of PR #161, but clean :) For simplicity purpose I copy-paste #161 here.
---------------------------
Hi,

I have been working on my own fork of tensorly for some time now, and I though I might start adding some contributions the easy way, with some small tweaks.

Here is a proposed solution to #72 (Fixed modes in Parafac). Since the optimization algorithm is ALS, it is quite straightforward to implement this. I simply changed the loop condition to ignore modes given by the user. I tested this on a more advanced branch of my fork, but this should do the trick.

Also, it is straightforward to to the same change to other functions such as nonnegative parafac or Tucker. I just wanted to try this simple PR first since I have little experience with it.

### CODE REVIEW

**Comment by JeanKossaifi** [comment]
Thanks @cohenjer!

Is `fixed_modes` meant to be a list of modes? If so, is the user meant to also pass in a list of factors? Otherwise we are just initializing the fixed modes randomly (or with SVD depending on `init`).

File: tensorly/decomposition/candecomp_parafac.py [hunk 1]
```diff
@@ -160,13 +163,17 @@ def parafac(tensor, rank, n_iter_max=100, init='svd', svd='numpy_svd',\
     weights = tl.ones(rank, **tl.context(tensor))
     Id = tl.eye(rank)*l2_reg
 
+    # MGR: fixed Modes
+    # Generating the mode update sequence
+    gen = (mode for mode in range(tl.ndim(tensor)) if mode not in fixed_modes)
```
Comment (JeanKossaifi) [inline tensorly/decomposition/candecomp_parafac.py]: Why a generator rather than a list?

**Comment by cohenjer** [comment]
- Yes fixed_modes would be a list of modes. And it makes indeed little sense to fix a mode if the initial factor values cannot be manually given. Arguably, this is a desirable feature anyway? I can add something like this in the PR, by modifying the initialize function and adding an optinal input.
- About generator vs list, I actually used lists in the other PR (which was more advanced). It does not really matter while we are not interested in picking an element in the generator. What is your opinion?

File: tensorly/decomposition/candecomp_parafac.py (hunk 1 above)
Comment (cohenjer) [reply tensorly/decomposition/candecomp_parafac.py]: I actually used lists in the other PR (which was more advanced). It does not really matter as long as we are not interested in picking an element in the generator. I can change that for sure.

File: tensorly/decomposition/candecomp_parafac.py [hunk 2]
```diff
@@ -15,8 +15,9 @@
 
 # License: BSD 3 clause
 
-def initialize_factors(tensor, rank, init='svd', svd='numpy_svd', random_state=None, 
-                       non_negative=False, normalize_factors=False):
+def initialize_factors(tensor, rank, init='svd', svd='numpy_svd', random_state=None,
```
Comment (cohenjer) [inline tensorly/decomposition/candecomp_parafac.py:18]: I chose to keep the parafac function as similar as possible with the original one, by changin instead the initialization function. Also this way, we can propagate the changes easily to other algorithms in the same file, and also to other problems such as Tucker which use a similar initialization function.

**Comment by JeanKossaifi** [comment]
Thanks @cohenjer, sorry for the delay.

Since we are adding this, I think it might be worth giving the option to give only a few input factors with their corresponding mode, e.g. 
`fixed_factors` being a list of tuples `(mode: int, factor : tensor)`.

**Comment by cohenjer** [comment]
So, I think fixing factors and providing initial factors are two different things. 
I agree it makes no sense (to me at least) to fix a mode and not provide an initial factor on that fixed mode. However, we might want to allow the user to use his own initialization, without fixing that mode.
Then having initial factors and fixed mode in one variable is a little weird. So I suggest that we keep fixed_modes and input_factors in two different variables.

We may provide a warning however when someone is fixing a mode but not providing an initial value.

What do you think ?

**Pull request closed without merging**